### PR TITLE
ZTransducer.contramap, mapError

### PR DESCRIPTION
**Pull request by regiskuckaertz**

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
uh ho.

**Comment by regiskuckaertz** [comment]
Weird timeout completely unrelated. Did something change with the generators?

**Comment by iravid** [comment]
Yes, I had to change some implementations in `Gen` and `Sample` to accomodate the ZStream API changes. Haven't had a chance to look into those yet

**Comment by iravid** [comment]
@regiskuckaertz I think I found the issue. Can you merge `feature/zconduit` into this branch?

File: streams-tests/shared/src/test/scala/zio/stream/ZTransducerSpec.scala [hunk 1]
```diff
@@ -6,6 +6,11 @@ import zio.test.Assertion._
 import zio.test._
 
 object ZTransducerSpec extends ZIOBaseSpec {
+  import ZIOTag._
+
+  val initErrorParser = ZTransducer.fromEffect(IO.fail("Ouch"))
```
Comment (iravid) [inline streams-tests/shared/src/test/scala/zio/stream/ZTransducerSpec.scala:11]: These are identical

File: streams-tests/shared/src/test/scala/zio/stream/ZTransducerSpec.scala [hunk 2]
```diff
@@ -18,7 +23,94 @@ object ZTransducerSpec extends ZIOBaseSpec {
     }
 
   def spec = suite("ZTransducerSpec")(
-    suite("Combinators")(),
+    suite("Combinators")(
+      suite("contramap")(
+        testM("happy path") {
```
Comment (iravid) [inline streams-tests/shared/src/test/scala/zio/stream/ZTransducerSpec.scala:28]: Can we cut down on some of these tests here and below? They're really just exercising `ZIO` at this point. Ideally we'd have property checks but not critical for now.

**Comment by regiskuckaertz** [comment]
Ha yeah, a semaphore 👀 Nice!

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 3]
```diff
@@ -642,21 +642,7 @@ abstract class ZStream[-R, +E, +O](
...
-                           case None    => Exit.succeed(None)
-                           case Some(e) => Exit.halt(e)
-                         },
-                         result => Exit.succeed(Some(result))
-                       )
-                     )
-
-                   }
-                 }
+                 .unfoldM(s)(s => f(s, left.pullElement, right.pullElement).flatMap(ZIO.done(_).optional))
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:645]: Nice!!!! We can probably apply this in a bunch of other places throughout the package :-)

**Review by iravid — APPROVED** [review]

**Pull request merged**